control-plane/piSerializer.cpp

```diff
@@ -0,0 +1,1471 @@
+  if (!expression->is<IR::Constant>()) {
+      ::error("Expected a constant for table size property: %1%", sizeProperty);
+      return defaultTableSize;
+  }
+
+  const int64_t tableSize = expression->to<IR::Constant>()->asInt();
+  return tableSize == 0 ? defaultTableSize : tableSize;
+}
+
+/// @return true iff @expression is a call to the 'isValid()' builtin.
+static const P4::BuiltInMethod*
+tryCastToIsValidBuiltin(const IR::Expression* expression,
+                        ReferenceMap* refMap,
+                        TypeMap* typeMap) {
+  if (!expression->is<IR::MethodCallExpression>()) return nullptr;
+  auto methodCall = expression->to<IR::MethodCallExpression>();
+  auto instance = P4::MethodInstance::resolve(methodCall, refMap, typeMap);
+  if (!instance->is<P4::BuiltInMethod>()) return nullptr;
+  auto builtin = instance->to<P4::BuiltInMethod>();
+  if (builtin->name != IR::Type_Header::isValid) return nullptr;
+  return builtin;
+}
+
+/// @return the header instance fields matched against by @table's key. The
+/// fields are represented as a (fully qualified field name, match type) tuple.
+static std::vector<std::pair<cstring, MatchType>>
+getMatchFields(const IR::P4Table* table, ReferenceMap* refMap, TypeMap* typeMap) {
+  std::vector<std::pair<cstring, MatchType>> matchFields;
+
+  auto key = table->getKey();
+  if (!key) return matchFields;
+
+  for (auto keyElement : *key->keyElements) {
+    auto matchTypeDecl = refMap->getDeclaration(keyElement->matchType->path, true)
+                               ->to<IR::Declaration_ID>();
+    BUG_CHECK(matchTypeDecl != nullptr, "No declaration for match type '%1%'",
+                                        keyElement->matchType);
+    const auto matchTypeName = matchTypeDecl->name.name;
+
+    auto expr = keyElement->expression;
+    if (expr->is<IR::Slice>()) {
+      expr = expr->to<IR::Slice>()->e0;
+    }
+
+    MatchType matchType;
+    if (matchTypeName == P4CoreLibrary::instance.exactMatch.name) {
+      matchType = MatchField::EXACT;
+
+      // If this is an exact match on the result of an isValid() call, we
+      // desugar that to a "valid" match.
+      auto callToIsValid = tryCastToIsValidBuiltin(expr, refMap, typeMap);
+      if (callToIsValid != nullptr) {
+        expr = callToIsValid->appliedTo;
+        matchType = MatchField::VALID;
+      }
+    } else if (matchTypeName == P4CoreLibrary::instance.lpmMatch.name) {
+      matchType = MatchField::LPM;
+    } else if (matchTypeName == P4CoreLibrary::instance.ternaryMatch.name) {
+      matchType = MatchField::TERNARY;
+
+      // If this is a ternary match on the result of an isValid() call, desugar
+      // that to a ternary match against the '$valid$' field of the appropriate
+      // header instance.
+      // XXX(seth): This is baking in BMV2-specific stuff. We should remove this
+      // by performing the desugaring in a separate compiler pass.
```

Comment:
Well this `$valid$` field is basically needed when a table is using `<hdr>.isValid()` as a match field. Because of the current p4info.proto, match fields have to be listed as `HeaderField` objects. I'm thinking in the longer term, if one gives a name annotation for a match field, one will expect that name to be used in the PI, not `<hdr>.$valid$`. We definitely need to modify p4info.proto to take into account name annotations for match fields. That can be seen as an orthogonal issue to adding `$valid$` to every header type though.